[AIRFLOW-5266] Allow aws_athena_hook to get all query results

lindsable opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5266#Athena.Client.get_query_results%5D) issues and references them in the PR title. 

### Description

- [x] The previous version of the hook returned up to 1000 results. The update allows the user to set the max results (default and max constrained by the api is 1000) and also the NextTokenId.  
- [x] The client request token in the run_query function is now truly optional. Boto3 doesn't allow you to send in parameters that are set to None.
- [x] Another function was added to return a paginator for the query results. 

### Tests

- [x] My PR adds the following unit tests:
tests/providers/amazon/aws/hooks/test_athena.py 

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

feluelle commented on airflow/contrib/hooks/aws_athena_hook.py:
```diff
@@ -157,6 +178,7 @@ def poll_query_status(self, query_execution_id, max_tries=None):
                 break
             if max_tries and try_number >= max_tries:  # Break loop if max_tries reached
                 final_query_state = query_state
+                self.stop_query(query_execution_id)
```
You should mention that in the commit :)

It fixes a bug, right?

feluelle commented on tests/contrib/hooks/test_aws_athena_hook.py:
```diff
@@ -0,0 +1,88 @@
+    AWSAthenaHook = None  # type: ignore
+
+
+class MockAthenaClient:
```
I am not a fan of this, because you have basically two `AthenaHook`s you need to maintain.
If you use `mock.patch` you can achieve the same without having two classes to keep in sync.

mik-laj replied:
I agree. We should use `unittest.mock` in this place.

lindsable replied:
I had an issue mocking this function because of the way boto3 implements the api. It's not a defined function but an api call. There are a few suggestions that I found [here:](https://stackoverflow.com/questions/37143597/mocking-boto3-s3-client-method-python) but we would still have to maintain it. The other aws hooks that use boto3 are able to use the [moto](https://github.com/spulec/moto) package.  but athena isn't mocked yet

OmerJog replied:
@mik-laj @feluelle  is there a solution for this ? Did some checks everyone everywhere complains it's impossible to mock athena with moto. How shall we proceeded?

feluelle replied:
You don't need the moto package to mock a class. Just mock the calls that are accessing the boto library in the get_query_results function. So that it will never really call the content of the moto get_query_results function.

feluelle replied:
Actually you only need to add `mock_conn.get_query_results.return_value = self.base_query_result` to the beginning of the test function. That is basically what you are already doing.

You are doing good actually. But I would remove the extra class.
This part
```
        if NextToken is None:
            return self.query_result_with_next
        else:
            return self.base_query_result
```
is logic that you don't need if you set the `mock_conn.get_query_results.return_value` directly in the test function.

A MockClass should not contain extra logic.

Do you understand what I mean? :D

lindsable replied:
Fixed!

ashb commented on airflow/providers/amazon/aws/hooks/athena.py:
```diff
@@ -133,7 +133,17 @@ def get_query_results(self, query_execution_id):
-        return self.get_conn().get_query_results(QueryExecutionId=query_execution_id)
+        result_params = {'QueryExecutionId': query_execution_id,
+                         'MaxResults': max_results}
+        results = self.get_conn().get_query_results(**result_params)
```
Does boto have a built-in pager for this?

ashb replied:
Yes, see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/athena.html#paginators and https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

lindsable replied:
Yes that does make sense.  Now that I think about it, it probably makes sense to just add a new function to check that the query is successful and then return the paginator. For the get_query_results function maybe we should just adding the option to send in the NextTokenId so you can paginate manually if you want.

ashb commented on tests/providers/amazon/aws/hooks/test_athena.py:
```diff
@@ -0,0 +1,120 @@
+from unittest import mock
+
+try:
+    from ..airflow.providers.amazon.aws.hooks.athena import AWSAthenaHook
```
`..` here is odd/unusual, please just use absolute import here

ashb commented on tests/providers/amazon/aws/hooks/test_athena.py:
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
```
Do we really not have any existing tests for the Athena hook? (i.e. Why is this a new file?)

lindsable replied:
That is correct.

feluelle commented on tests/providers/amazon/aws/hooks/test_athena.py:
```diff
@@ -0,0 +1,120 @@
+MOCK_GET_QUERY_EXECUTION = {'QueryExecution': {'Status': {'State': 'SUCCEEDED'}}}
+MOCK_ATHENA_QUERY_ID = 'MockQueryExecutionId'
+
+# All other functions are tested in test/contrib/operators/test_aws_athena_operator.py
```
```suggestion
```

feluelle commented on airflow/providers/amazon/aws/hooks/athena.py:
```diff
@@ -116,7 +120,7 @@ def get_state_change_reason(self, query_execution_id):
             return reason  # pylint: disable=lost-exception
 
-    def get_query_results(self, query_execution_id):
+    def get_query_results(self, query_execution_id, next_token_id=None, max_results=1000):
```
Please add docs regarding next_token_id and max_results

feluelle commented on airflow/providers/amazon/aws/hooks/athena.py:
```diff
@@ -132,7 +136,37 @@ def get_query_results(self, query_execution_id):
+            result_params['NextToken'] = next_token_id
+        return self.get_conn().get_query_results(**result_params)
+
+    def get_query_results_paginator(self, query_execution_id, max_results=1000):
```
Same here.

ashb commented on airflow/providers/amazon/aws/hooks/athena.py:
```diff
@@ -163,6 +213,7 @@ def poll_query_status(self, query_execution_id, max_tries=None):
                 break
             if max_tries and try_number >= max_tries:  # Break loop if max_tries reached
                 final_query_state = query_state
+                self.stop_query(query_execution_id)
```
This line appears to be unrelated to fetching all results or not, right?

I guess this is probably a sane default, but this is a change in the current behaviour. Previously any query would continue running meaning that even if it takes a lot longer in Airflow it would (eventually) be available in the Athena console. 

Hmmm.

ashb replied:
Other than this (which I don't know if we want or not) the code looks good, and tests are always great :)

ashb replied:
@lindsable What do you think? Is this required? Is it a sane default? Are there cases where users might want the query to continue anyway?

lindsable replied:
While I can't think of anything specific, I think you have a good point about changing existing functionality. We will leave it up to the end user to stop query if they choose. I'll remove this line.

ashb replied:
Can you ping me once this has been changed?

(Sorry this ticket has sat for _so_ long.)

lindsable replied:
Changed! Just waiting for the tests to pass again. @ashb

lindsable replied:
@ashb I'm not sure why the static checks are failing on the docker lint. Any insights?

ashb replied:
@lindsable Yeah does seem unrelated to your change -- try rebasing on to latest master.

lindsable replied:
Looks like the same tests are also failing on master :(

ashb reviewed (commented):
Sorry it took rather more than the promised "tomorrow".

ashb commented on airflow/providers/amazon/aws/hooks/athena.py:
```diff
@@ -125,7 +133,49 @@ def get_query_results(self, query_execution_id):
+        """
+        query_state = self.check_query_status(query_execution_id)
+        if query_state is None:
+            self.log.error('Invalid Query state')
```
```suggestion
            self.log.error('Invalid Query state (null)')
```

ashb commented on airflow/providers/amazon/aws/hooks/athena.py:
```diff
@@ -125,7 +133,49 @@ def get_query_results(self, query_execution_id):
+        if query_state is None:
+            self.log.error('Invalid Query state')
+            return None
+        elif query_state in self.INTERMEDIATE_STATES or query_state in self.FAILURE_STATES:
```
```suggestion
        if query_state in self.INTERMEDIATE_STATES or query_state in self.FAILURE_STATES:
```

ashb commented on tests/providers/amazon/aws/hooks/test_athena.py:
```diff
@@ -0,0 +1,172 @@
+class TestAWSAthenaHook(unittest.TestCase):
+
+    def setUp(self):
+        self.athena = AWSAthenaHook(sleep_time=1)
```
Can we do sleep time=0 instead? (`sleep(0)` is valid) - This would probably make the tests run quicker.

ashb reviewed (changes requested):
A few tiny changes, this is looking good.

Sorry it's been open for so very long!

dwjoss committed:
Improve AWSAthenaHook

make client request token optional

add optional params NextToken and MaxResults to get_query_results

add get_query_results_paginator

Stop query execution if it times out

add tests for AWSAthenaHook

remove 'tests/providers/amazon/aws/hooks/test_athena.py' from missing tests

add docs for new parameters

update tests

Don't stop query if polling reaches max tries

small improvements
